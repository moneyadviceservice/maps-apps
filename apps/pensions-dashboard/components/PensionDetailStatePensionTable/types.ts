export type Row = {
  title: string;
  monthlyAmount: string;
  yearlyAmount: string;
  payableDate: string;
};
