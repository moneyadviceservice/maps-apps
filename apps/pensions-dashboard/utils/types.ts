export type ErrorResponse = {
  error: string;
  message: string;
};

export type PensionData = {
  pensionPolicies: PensionPolicy[];
  peiInformation: PeiInformationModel;
  pensionsDataRetrievalComplete: boolean;
};

type PensionPolicy = {
  pensionArrangements: PensionArrangement[];
};

export type PensionArrangement = {
  externalPensionPolicyId: string;
  externalAssetId: string;
  matchType: 'DEFN' | 'POSS' | 'CONT' | 'NEW' | 'SYS';
  schemeName: string;
  alternateSchemeNames?: AlternateScheme[];
  contactReference: string;
  pensionType: 'AVC' | 'CB' | 'CDC' | 'DB' | 'DC' | 'HYB' | 'SP';
  pensionOrigin?: 'A' | 'PC' | 'PM' | 'PT' | 'WC' | 'WM' | 'WT';
  pensionStatus?: 'A' | 'I' | 'IPPF' | 'IWU';
  startDate?: string;
  retirementDate?: string;
  dateOfBirth?: string;
  statePensionMessageEng?: string;
  statePensionMessageWelsh?: string;
  contributionsFromMultipleEmployers: boolean;
  pensionAdministrator: PensionAdministrator;
  employmentMembershipPeriods?: EmploymentMembershipPeriod[];
  benefitIllustrations?: BenefitIllustration[];
  additionalDataSources?: AdditionalDataSource[];
};

type AlternateScheme = {
  name: string;
  alternateNameType: 'FOR' | 'OTH';
};

type PensionAdministrator = {
  name: string;
  contactMethods: ContactMethod[];
};

type ContactMethod = {
  preferred: boolean;
  contactMethodDetails: EmailAddress | PhoneNumber | PostalAddress | Website;
};

type EmploymentMembershipPeriod = {
  employerName: string;
  employerStatus: 'C' | 'H';
  membershipStartDate: string;
  membershipEndDate: string;
};

type BenefitIllustration = {
  illustrationComponents: BenefitIllustrationComponent[];
  illustrationDate: string;
};

type AdditionalDataSource = {
  url: string;
  informationType: 'C_AND_C' | 'SIP' | 'IMP' | 'ANR' | 'SP';
};

type PeiInformationModel = {
  peiRetrievalComplete: boolean;
  peiData: PeiData[];
};

type PeiData = {
  pei: string;
  description: string;
  retrievalStatus:
    | 'NEW'
    | 'RETRIEVAL_REQUESTED'
    | 'RETRIEVAL_COMPLETE'
    | 'RETRIEVAL_FAILED';
};

type EmailAddress = {
  email: string;
};

type PhoneNumber = {
  number: string;
  usage: ('A' | 'M' | 'N' | 'S' | 'W')[];
};

type PostalAddress = {
  postalName: string;
  line1: string;
  line2?: string;
  line3?: string;
  line4?: string;
  line5?: string;
  postcode: string;
  countryCode: string;
};

type Website = {
  url: string;
};

type BenefitIllustrationComponent = {
  illustrationType: 'ERI' | 'AP';
  unavailableReason?:
    | 'ANO'
    | 'DB'
    | 'DBC'
    | 'DCC'
    | 'DCHA'
    | 'DCHP'
    | 'DCSM'
    | 'MEM'
    | 'NEW'
    | 'NET'
    | 'PPF'
    | 'TRN'
    | 'WU';
  benefitType: 'AVC' | 'CBL' | 'CBS' | 'CDI' | 'CDL' | 'DB' | 'DBL' | 'DC';
  calculationMethod: 'BS' | 'CBI' | 'SMPI';
  payableDetails: PayableDetails;
  dcPot?: number;
  survivorBenefit?: boolean;
  safeguardedBenefit?: boolean;
  illustratonWarnings?: (
    | 'AVC'
    | 'CUR'
    | 'DEF'
    | 'FAS'
    | 'PEO'
    | 'PNR'
    | 'PSO'
    | 'SPA'
    | 'TVI'
    | 'INP'
  )[];
};

type PayableDetails =
  | RecurringIncomeDetails
  | LumpSumDetails
  | AmountNotProvidedDetails;

type RecurringIncomeDetails = {
  amountType: 'INC' | 'INCL' | 'INCN';
  annualAmount: number;
  monthlyAmount: number;
  lastPaymentDate?: string;
  payableDate: string;
  increasing: boolean;
};

type LumpSumDetails = {
  amountType: 'CSH' | 'CSHL' | 'CSHN';
  amount: number;
  payableDate: string;
};

type AmountNotProvidedDetails = {
  reason: 'SML';
  payableDate: string;
  lastPaymentDate?: string;
};

export type PensionsOverviewModel = {
  totalPensions: number;
  confirmedPensions: PensionArrangement[];
  unconfirmedPensions: PensionArrangement[];
  incompletePensions: PensionArrangement[];
};
