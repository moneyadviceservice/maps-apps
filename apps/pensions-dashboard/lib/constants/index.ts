import Cookies from 'cookies';

export const STATE_RETIREMENT_AGE = 67;

export const NO_DATA = '--';

export const NUMBER_WORDS = [
  { en: 'zero', cy: 'sero' },
  { en: 'one', cy: 'un' },
  { en: 'two', cy: 'dau' },
  { en: 'three', cy: 'tri' },
  { en: 'four', cy: 'pedwar' },
  { en: 'five', cy: 'pump' },
  { en: 'six', cy: 'chwech' },
  { en: 'seven', cy: 'saith' },
  { en: 'eight', cy: 'wyth' },
  { en: 'nine', cy: 'naw' },
  { en: 'ten', cy: 'deg' },
  { en: 'eleven', cy: 'un ar ddeg' },
  { en: 'twelve', cy: 'deuddeg' },
];

export const PROTOCOL =
  process?.env.NODE_ENV === 'development' ? 'http://' : 'https://';

export const COOKIE_OPTIONS = {
  path: '/',
  httpOnly: true,
  sameSite: 'lax',
} as Cookies.SetOption;

export enum PensionsCardType {
  CONFIRMED = 'CONFIRMED',
  CONFIRMED_NO_INCOME = 'CONFIRMED_NO_INCOME',
  PENDING = 'PENDING',
}

export enum MatchType {
  DEFN = 'DEFN',
  POSS = 'POSS',
  CONT = 'CONT',
  NEW = 'NEW',
  SYS = 'SYS',
}

export enum PensionType {
  AVC = 'AVC',
  CB = 'CB',
  CDC = 'CDC',
  DB = 'DB',
  DC = 'DC',
  HYB = 'HYB',
  SP = 'SP',
}

export enum PensionOrigin {
  A = 'A',
  PC = 'PC',
  PM = 'PM',
  PT = 'PT',
  WC = 'WC',
  WM = 'WM',
  WT = 'WT',
}

export enum PensionStatus {
  A = 'A',
  I = 'I',
  IPPF = 'IPPF',
  IWU = 'IWU',
}

export enum AlternateNameType {
  FOR = 'FOR',
  OTH = 'OTH',
}

export enum EmployerStatus {
  C = 'C',
  H = 'H',
}

export enum InformationType {
  C_AND_C = 'C_AND_C',
  SIP = 'SIP',
  IMP = 'IMP',
  ANR = 'ANR',
  SP = 'SP',
}

export enum RetrievalStatus {
  NEW = 'NEW',
  RETRIEVAL_REQUESTED = 'RETRIEVAL_REQUESTED',
  RETRIEVAL_COMPLETE = 'RETRIEVAL_COMPLETE',
  RETRIEVAL_FAILED = 'RETRIEVAL_FAILED',
}

export enum UsageType {
  A = 'A',
  M = 'M',
  N = 'N',
  S = 'S',
  W = 'W',
}

export enum IllustrationType {
  ERI = 'ERI',
  AP = 'AP',
  DB = 'DB',
}

export enum BenefitType {
  AVC = 'AVC',
  CBL = 'CBL',
  CBS = 'CBS',
  CDI = 'CDI',
  CDL = 'CDL',
  DB = 'DB',
  DBL = 'DBL',
  DC = 'DC',
  SP = 'SP',
}

export enum UnavailableReason {
  ANO = 'ANO',
  DB = 'DB',
  DBC = 'DBC',
  DCC = 'DCC',
  DCHA = 'DCHA',
  DCHP = 'DCHP',
  DCSM = 'DCSM',
  MEM = 'MEM',
  NEW = 'NEW',
  NET = 'NET',
  PPF = 'PPF',
  TRN = 'TRN',
  WU = 'WU',
}

export enum IllustrationWarning {
  AVC = 'AVC',
  CUR = 'CUR',
  DEF = 'DEF',
  FAS = 'FAS',
  PEO = 'PEO',
  PNR = 'PNR',
  PSO = 'PSO',
  SPA = 'SPA',
  TVI = 'TVI',
  INP = 'INP',
}

export enum CalculationMethod {
  BS = 'BS',
  CBI = 'CBI',
  SMPI = 'SMPI',
}

export enum LumpSumAmountType {
  CSH = 'CSH',
  CSHL = 'CSHL',
  CSHN = 'CSHN',
}

export enum RecurringAmountType {
  INC = 'INC',
  INCL = 'INCL',
  INCN = 'INCN',
}

export enum AmountNotProvidedReason {
  SML = 'SML',
}

export enum ContactMethods {
  TELEPHONE = 'number',
  EMAIL = 'email',
  POSTAL_NAME = 'postalName',
  WEB_CONTACTS = 'url',
}

export const BACK_LINKS: { [key in string]: string } = {
  CONFIRMED: '/your-pension-breakdown',
  INCOMPLETE: '/pending-pensions',
  UNCONFIRMED: '/pensions-that-need-action',
  '': '/your-pension-search-results',
};

// API response messages
export const REQUEST_ABANDONED = 'Network request was abandoned';
export const RESPONSE_NOT_OK = 'Network response was not ok';
export const RESPONSE_NOT_ACCEPTED = 'Network response was not accepted';
export const REQUEST_FAILED = 'Network request failed';
export const DATA_NOT_FOUND = 'Data not found';

export const DEFAULT_LOAD_TIME = 80;

export enum PensionGroup {
  GREEN = 'green',
  GREEN_NO_INCOME = 'green_no_income',
  RED = 'red',
  YELLOW = 'yellow',
}
