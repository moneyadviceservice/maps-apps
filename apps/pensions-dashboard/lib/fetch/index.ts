export * from './delete-pension-data';
export * from './get-all-pensions';
export * from './get-claims-gathering-redirect';
export * from './get-csrf-token';
export * from './get-load-times';
export * from './get-pension-detail';
export * from './get-pensions-data';
export * from './get-pensions-overview';
export * from './post-pensions-data';
export * from './post-pensions-data-retrieval';
export * from './post-redirect-details';
export * from './validate-access-token';
