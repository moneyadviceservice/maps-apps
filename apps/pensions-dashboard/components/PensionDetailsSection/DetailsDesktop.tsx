import { twMerge } from 'tailwind-merge';

import useTranslation from '@maps-react/hooks/useTranslation';

import { PensionType } from '../../lib/constants';
import { DetailRow } from '../PensionDetailRow';
import { TableSection } from '../TableSection';
import { Row } from './PensionDetailsSection';

const tdClasses = 'pb-3 md:pt-[14px] md:pb-4 text-left align-top max-md:block';
const thClasses =
  'text-left align-top font-bold pt-[10px] pb-[2px] md:pt-[14px] md:pb-4 max-md:block border-b-1 border-b-slate-400';

type DetailsDesktopProps = {
  data: Row[];
  type: PensionType;
};

export const DetailsDesktop = ({ data, type }: DetailsDesktopProps) => {
  const { t } = useTranslation();

  return (
    <div className="hidden md:block">
      <TableSection heading={t('pages.pension-details.details.title')}>
        <thead>
          <tr>
            <th className={twMerge(thClasses, 'md:w-1/3 lg:w-2/5')}></th>
            <th className={twMerge(thClasses, '')}>
              {t(
                `pages.pension-details.details.${
                  type === PensionType.SP ? 'estimate-today' : 'current-value'
                }`,
              )}
            </th>
            <th className={twMerge(thClasses, '')}>
              {t(
                `pages.pension-details.details.${
                  type === PensionType.SP
                    ? 'forecast'
                    : 'estimate-at-retirement'
                }`,
              )}
            </th>
          </tr>
        </thead>

        <tbody className="max-md:block">
          {data.map(({ title, currentValue, retirementValue }) => (
            <DetailRow
              key={title}
              heading={t(`pages.pension-details.details.${title}`)}
            >
              <td className={tdClasses} data-testid={`current-${title}`}>
                {currentValue}
              </td>
              <td className={tdClasses} data-testid={`retirement-${title}`}>
                {retirementValue}
              </td>
            </DetailRow>
          ))}
        </tbody>
      </TableSection>
    </div>
  );
};
