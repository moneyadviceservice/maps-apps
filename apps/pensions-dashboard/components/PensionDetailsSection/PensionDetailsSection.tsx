import { Heading } from '@maps-react/common/components/Heading';
import useTranslation from '@maps-react/hooks/useTranslation';

import { PensionArrangement } from '../../lib/types';
import { BenefitsIllustrationsTable } from '../BenefitIllustrationsTable/BenefitIllustrationsTable';

type PensionDetailMoreInfoProps = {
  data: PensionArrangement;
  hasPayableDetails: boolean;
};

export const PensionDetailsSection = ({
  data,
  hasPayableDetails,
}: PensionDetailMoreInfoProps) => {
  const { t } = useTranslation();

  if (!data.benefitIllustrations) {
    return null;
  }

  return (
    <>
      <Heading level="h2">{t('pages.pension-details.details.heading')}</Heading>
      <BenefitsIllustrationsTable
        data={data.benefitIllustrations}
        hasPayableDetails={hasPayableDetails}
      />
    </>
  );
};
