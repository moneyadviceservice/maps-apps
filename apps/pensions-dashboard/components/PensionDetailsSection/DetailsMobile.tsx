import { Heading } from '@maps-react/common/components/Heading';
import useTranslation from '@maps-react/hooks/useTranslation';

import { PensionType } from '../../lib/constants';
import { Row } from './PensionDetailsSection';

type DetailsMobileProps = {
  data: Row[];
  type: PensionType;
};

export const DetailsMobile = ({ data, type }: DetailsMobileProps) => {
  const { t } = useTranslation();

  return (
    <div className="md:hidden mt-14 md:mt-16">
      <Heading
        data-testid="table-section-heading"
        level="h2"
        className="max-md:mb-4 md:text-5xl md:mb-4"
      >
        {t('pages.pension-details.details.title')}
      </Heading>

      <div className="max-md:mb-8 *:border-b-1 *:border-b-slate-400 *:py-2">
        <Heading level="h3" className="font-semibold">
          {t(
            `pages.pension-details.details.${
              type === PensionType.SP ? 'estimate-today' : 'mobile-current'
            }`,
          )}
        </Heading>

        {data.map(({ title, currentValue }) => (
          <div key={title}>
            <div className="font-bold">
              {t(`pages.pension-details.details.${title}`)}
            </div>
            <div data-testid={`current-${title}-mobile`}>{currentValue}</div>
          </div>
        ))}

        <Heading level="h3" className="mt-8 font-semibold">
          {t(
            `pages.pension-details.details.${
              type === PensionType.SP ? 'forecast' : 'mobile-retirement'
            }`,
          )}
        </Heading>

        {data.map(({ title, retirementValue }) => (
          <div key={title}>
            <div className="font-bold">
              {t(`pages.pension-details.details.${title}`)}
            </div>
            <div data-testid={`retirement-${title}-mobile`}>
              {retirementValue}
            </div>
          </div>
        ))}
      </div>
    </div>
  );
};
