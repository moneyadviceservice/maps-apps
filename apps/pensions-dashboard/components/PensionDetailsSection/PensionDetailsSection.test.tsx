import { render } from '@testing-library/react';

import useTranslation from '@maps-react/hooks/useTranslation';

import { mockPensionsData } from '../../lib/mocks';
import { PensionArrangement } from '../../lib/types';
import { PensionDetailsSection } from './PensionDetailsSection';

import '@testing-library/jest-dom/extend-expect';

jest.mock('@maps-react/hooks/useTranslation');

const mockUseTranslation = useTranslation as jest.Mock;

const mockData = mockPensionsData.pensionPolicies[0]
  .pensionArrangements[0] as PensionArrangement;

describe('PensionDetailsSection', () => {
  beforeEach(() => {
    mockUseTranslation.mockReturnValue({
      t: (key: string) => key,
    });
  });

  it('renders correctly', () => {
    const { container } = render(
      <PensionDetailsSection data={mockData} hasPayableDetails={true} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders heading and table when benefitIllustrations exist', () => {
    const { getByText } = render(
      <PensionDetailsSection data={mockData} hasPayableDetails={true} />,
    );

    expect(
      getByText('pages.pension-details.details.heading'),
    ).toBeInTheDocument();
    expect(
      getByText('pages.pension-details.details.heading'),
    ).toBeInTheDocument();
  });

  it('does not render when benefitIllustrations do not exist', () => {
    const data: PensionArrangement = {} as PensionArrangement;

    const { container } = render(
      <PensionDetailsSection data={data} hasPayableDetails={true} />,
    );

    expect(container.firstChild).toBeNull();
  });

  it('renders table with correct payable details', () => {
    const { getByText, queryAllByText } = render(
      <PensionDetailsSection data={mockData} hasPayableDetails={false} />,
    );

    expect(getByText('ERI')).toBeInTheDocument();
    expect(getByText('AP')).toBeInTheDocument();
    expect(
      getByText('pages.pension-details.details.benefit-type'),
    ).toBeInTheDocument();
    expect(getByText('pages.pension-details.headings.pot')).toBeInTheDocument();
    expect(queryAllByText('Â£540,500').length).toBeGreaterThan(0);
    expect(queryAllByText('Defined Contribution').length).toBeGreaterThan(0);
  });
});
