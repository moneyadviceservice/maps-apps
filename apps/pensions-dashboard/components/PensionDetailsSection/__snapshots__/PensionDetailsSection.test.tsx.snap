// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PensionDetailsSection renders correctly 1`] = `
<div>
  <h2
    class="text-3xl md:text-4xl text-gray-800 font-bold"
  >
    pages.pension-details.details.heading
  </h2>
  <div
    class="relative mb-16 overflow-x-auto"
  >
    <table
      class="w-[800px] md:w-full"
    >
      <thead>
        <tr
          class="border-b-1 border-b-slate-400"
        >
          <th
            class="w-1/4 md:w-1/2 border-b-1 border-b-slate-400"
          />
          <th
            class="py-3 text-left align-top"
          >
            ERI
          </th>
          <th
            class="py-3 text-left align-top"
          >
            AP
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="border-b-1 border-b-slate-400"
        >
          <td
            class="py-3 text-left align-top font-bold"
          >
            pages.pension-details.details.benefit-type
          </td>
          <td
            class="py-3 text-left align-top"
          >
            Defined Contribution
          </td>
          <td
            class="py-3 text-left align-top"
          >
            Defined Contribution
          </td>
        </tr>
        <tr
          class="border-b-1 border-b-slate-400"
        >
          <td
            class="py-3 text-left align-top font-bold"
          >
            pages.pension-details.details.annual-amount
          </td>
          <td
            class="py-3 text-left align-top"
          >
            common.amount-unavailable
             
            common.a-year
          </td>
          <td
            class="py-3 text-left align-top"
          >
            common.amount-unavailable
             
            common.a-year
          </td>
        </tr>
        <tr
          class="border-b-1 border-b-slate-400"
        >
          <td
            class="py-3 text-left align-top font-bold"
          >
            pages.pension-details.details.monthly-amount
          </td>
          <td
            class="py-3 text-left align-top"
          >
            common.amount-unavailable
             
            common.a-month
          </td>
          <td
            class="py-3 text-left align-top"
          >
            common.amount-unavailable
             
            common.a-month
          </td>
        </tr>
        <tr
          class="border-b-1 border-b-slate-400"
        >
          <td
            class="py-3 text-left align-top font-bold"
          >
            pages.pension-details.details.payable-date
          </td>
          <td
            class="py-3 text-left align-top"
          >
            date unavailable
          </td>
          <td
            class="py-3 text-left align-top"
          >
            date unavailable
          </td>
        </tr>
        <tr
          class="border-b-1 border-b-slate-400"
        >
          <td
            class="py-3 text-left align-top font-bold"
          >
            pages.pension-details.headings.pot
          </td>
          <td
            class="py-3 text-left align-top"
          >
            £540,500
          </td>
          <td
            class="py-3 text-left align-top"
          >
            £540,500
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
